cmd/kaspawallet/libkaspawallet/bip32/serialization.go

```diff
@@ -0,0 +1,142 @@
+	)
+
+	isPrivate := isPrivateVersion(extKey.Version)
+	if isPrivate {
+		privateKeyPadding := serialized[versionSerializationLen+
+			depthSerializationLen+
+			fingerprintSerializationLen+
+			childNumberSerializationLen+
+			chainCodeSerializationLen]
+		if privateKeyPadding != 0 {
+			return nil, errors.Errorf("expected 0 padding for private key but got %d", privateKeyPadding)
+		}
+
+		extKey.PrivateKey, err = secp256k1.DeserializeECDSAPrivateKeyFromSlice(serialized[versionSerializationLen+
+			depthSerializationLen+
+			fingerprintSerializationLen+
+			childNumberSerializationLen+
+			chainCodeSerializationLen+1 : versionSerializationLen+
+			depthSerializationLen+
+			fingerprintSerializationLen+
+			childNumberSerializationLen+
+			chainCodeSerializationLen+
+			keySerializationLen])
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		extKey.publicKey, err = secp256k1.DeserializeECDSAPubKey(serialized[versionSerializationLen+
+			depthSerializationLen+
+			fingerprintSerializationLen+
+			childNumberSerializationLen+
+			chainCodeSerializationLen : versionSerializationLen+
+			depthSerializationLen+
+			fingerprintSerializationLen+
+			childNumberSerializationLen+
+			chainCodeSerializationLen+
+			keySerializationLen])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return extKey, nil
+}
+
+func (extKey *ExtendedKey) serialize() ([]byte, error) {
+	var serialized [extendedKeySerializationLen]byte
+	copy(serialized[:versionSerializationLen], extKey.Version[:])
+	serialized[versionSerializationLen] = extKey.Depth
+	copy(serialized[versionSerializationLen+depthSerializationLen:], extKey.Fingerprint[:])
+	binary.BigEndian.PutUint32(
+		serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen:],
+		extKey.ChildNumber,
+	)
+	copy(
+		serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen+childNumberSerializationLen:],
+		extKey.ChainCode[:],
+	)
+
+	if extKey.IsPrivate() {
+		serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen+childNumberSerializationLen+chainCodeSerializationLen] = 0
+		copy(
+			serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen+childNumberSerializationLen+chainCodeSerializationLen+1:],
+			extKey.PrivateKey.Serialize()[:],
+		)
+	} else {
+		publicKey, err := extKey.PublicKey()
+		if err != nil {
+			return nil, err
+		}
+
+		serializedPublicKey, err := publicKey.Serialize()
+		if err != nil {
+			return nil, err
+		}
+
+		copy(
+			serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen+childNumberSerializationLen+chainCodeSerializationLen:],
+			serializedPublicKey[:],
+		)
+	}
+
+	checkSum := doubleSha256(serialized[:len(serialized)-checkSumLen])
+	copy(
+		serialized[versionSerializationLen+depthSerializationLen+fingerprintSerializationLen+childNumberSerializationLen+chainCodeSerializationLen+keySerializationLen:],
+		checkSum,
+	)
+	return serialized[:], nil
```

Comment:
nit, I think the checksum is more part of the base58 encoding than part of the serialization,
but `serialize` is currently a private function so we don't care about the distinction (it also saves use from another allocation+copy)